src-control/Control/RefCount.hs

```diff
@@ -131,3 +165,231 @@ readRefCount RefCounter{countVar} = RefCount <$> readPrimVar countVar
+-- reference).
+--
+releaseRef :: (HasFinaliser obj, FinaliserM obj ~ m, PrimMonad m)
+           => Ref obj -> m ()
+#ifndef NO_IGNORE_ASSERTS
+releaseRef (Ref obj) = unsharedFinaliser obj
+#else
+releaseRef (Ref obj released) = do
+    assertNotReleased released
+    unsharedFinaliser obj
+    unsafeIOToPrim $ writeIORef released True
+#endif
+
+-- | Use the object in a 'Ref'. Do not retain the object after the scope of
+-- the body. If you cannot use scoped \"with\" style, use 'unsafeDeRef'.
+--
+#ifndef NO_IGNORE_ASSERTS
+withRef :: Ref obj -> (obj -> m a) -> m a
+withRef (Ref obj) f = f obj
+#else
+withRef :: PrimMonad m => Ref obj -> (obj -> m a) -> m a
+withRef (Ref obj released) f = assertNotReleased released
+                            >> f obj
+#endif
+
+-- | Get the object in a 'Ref'. Be careful with retaining the object for too
+-- long, since the object must not be used after 'releaseRef' is called.
+--
+unsafeDeRef :: Ref obj -> obj
+#ifndef NO_IGNORE_ASSERTS
+unsafeDeRef (Ref obj) = obj
+#else
+unsafeDeRef (Ref obj released) =
+          unsafeDupablePerformIO (assertNotReleased released)
+    `seq` obj
+#endif
+
+-- | Duplicate an existing reference, to produce a new reference.
+--
+dupRef :: (HasSharedFinaliser obj, FinaliserM obj ~ m, PrimMonad m)
+       => Ref obj -> m (Ref obj)
+#ifndef NO_IGNORE_ASSERTS
+dupRef (Ref obj) = do
+    incrementRefCounter (getRefCounter obj)
+    return (Ref obj)
+#else
+dupRef (Ref obj released) = do
+    assertNotReleased released
+    incrementRefCounter (getRefCounter obj)
+    released' <- unsafeIOToPrim $ newIORef False
+    return (Ref obj released')
+#endif
+
+
+-- | A \"weak\" reference to an object: that is, a reference that does not
+-- guarantee to keep the object alive. If however the object is still alive
+-- (due to other normal references still existing) then it can be converted
+-- back into a normal reference with 'deRefWeak'.
+--
+-- Weak references do not themselves need to be released.
+--
+newtype WeakRef a = WeakRef a
+
+-- | Given an existing normal reference, create a new weak reference.
+--
+mkWeakRef :: Ref obj -> WeakRef obj
+#ifndef NO_IGNORE_ASSERTS
+mkWeakRef (Ref obj) = WeakRef obj
+#else
+mkWeakRef (Ref obj _) = WeakRef obj
+#endif
+
+-- | If the object is still alive, obtain a /new/ normal reference. The normal
+-- rules for 'Ref' apply, including the need to eventually calling 'releaseRef'.
+--
+deRefWeak :: (HasSharedFinaliser obj, FinaliserM obj ~ m, PrimMonad m)
+          => WeakRef obj -> m (Maybe (Ref obj))
+#ifndef NO_IGNORE_ASSERTS
+deRefWeak (WeakRef obj) = do
+    success <- tryIncrementRefCounter (getRefCounter obj)
+    if success then return (Just (Ref obj))
+               else return Nothing
+#else
+deRefWeak (WeakRef obj) = do
+    success <- tryIncrementRefCounter (getRefCounter obj)
+    if success
+      then do released <- unsafeIOToPrim $ newIORef False
+              return (Just (Ref obj released))
+      else return Nothing
+#endif
```

Comment:
This CPP could all go away using `newRef`, though, right?

Reply:
Not any more. `newRef` makes a new ref counter, here we need it shared.